x-pack/plugins/lens/public/xy_visualization/axis_settings_popover.tsx

```diff
@@ -0,0 +1,244 @@
+  layers,
+  popoverTitle,
+  axis,
+  axisTitle,
+  frame,
+  updateTitleState,
+  toggleTickLabelsVisibility,
+  toggleGridlinesVisibility,
+  isDisabled,
+  areTickLabelsVisible,
+  areGridlinesVisible,
+  isAxisTitleVisible,
+  toggleAxisTitleVisibility,
+}) => {
+  const [popoversOpenState, setPopoversOpenState] = useState(false);
+  const [title, setTitle] = useState<string | undefined>(axisTitle);
+
+  const getAxisTitle = useCallback(() => {
+    const defaultTitle = axisTitle;
+
+    const initialLayer = layers?.[0];
+    if (!initialLayer || !initialLayer.accessors.length) {
+      return defaultTitle;
+    }
+    const datasource = frame.datasourceLayers[initialLayer.layerId];
+    if (!datasource) {
+      return defaultTitle;
+    }
+    switch (axis) {
+      case 'yLeft':
+        const yLeft = initialLayer.accessors[0]
+          ? datasource.getOperationForColumnId(initialLayer.accessors[0])
+          : null;
+        return defaultTitle || yLeft?.label;
+      case 'x':
+      default:
+        const x = initialLayer.xAccessor
+          ? datasource.getOperationForColumnId(initialLayer.xAccessor)
+          : null;
+        return defaultTitle || x?.label;
+      case 'yRight':
+        let yRightTitle: string = '';
+        const rightAxisLayer = layers?.filter(
+          (layer) =>
+            layer.yConfig && layer.yConfig.length > 0 && layer.yConfig[0].axisMode === 'right'
+        );
+        if (rightAxisLayer && rightAxisLayer.length > 0) {
+          const yConfig = rightAxisLayer[0].yConfig;
+          const dataSourceNewLayer = frame.datasourceLayers[rightAxisLayer[0].layerId];
+          const y = yConfig
+            ? dataSourceNewLayer.getOperationForColumnId(yConfig[0].forAccessor)
+            : null;
+          yRightTitle = y?.label || '';
+        }
+        return defaultTitle || yRightTitle;
+    }
+    /* We want this callback to run only if open changes its state. What we want to accomplish here is to give the user a better UX.
+       By default these input fields have the axis legends. If the user changes the input text, the axis legends should also change.
+       BUT if the user cleans up the input text, it should remain empty until the user closes and reopens the panel.
+       In that case, the default axis legend should appear. */
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [popoversOpenState]);
+
+  useEffect(() => {
+    setTitle(getAxisTitle());
+  }, [getAxisTitle]);
+
+  const onTitleChange = (value: string): void => {
+    setTitle(value);
+    updateTitleState(value);
+  };
+  return (
+    <ToolbarPopover
+      title={popoverTitle}
+      handlePopoverState={setPopoversOpenState}
+      type={popoverConfig(axis).icon}
+      groupPosition={popoverConfig(axis).groupPosition}
+      isDisabled={isDisabled}
+    >
+      <EuiFlexGroup gutterSize="s" justifyContent="spaceBetween">
+        <EuiFlexItem grow={false}>
+          <EuiText size="xs">
+            <h4>
```

Comment:
Nitpick: I haven't checked this, but I know that Lens has issues with the [page heading structure](https://github.com/elastic/kibana/issues/37539). This use of `h4` might be problematic, or might be correct, but maybe you can check?

Reply:
Thanx for this tip! It seems to be ok 
<img width="150" alt="Screenshot 2020-09-09 at 10 33 07 AM" src="https://user-images.githubusercontent.com/17003240/92568451-df5ad480-f287-11ea-9a4c-4f90cd957835.png">